### ISSUE TITLE
Controller fails to delete VM's in failed state

### ISSUE THREAD

**Issue by desek** [kubernetes-sigs/cluster-api-provider-azure#2571]
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**

1. Created a new `MachineDeployment` with an `AzureMachineTemplate` that uses spot instances
2. The Azure VM creation fails with Azure error `OverconstrainedZonalAllocationRequest` and set VM in failed state
3. The capz-controller tries to delete the VM and gets error message:
```
I0818 12:34:33.215446       1 recorder.go:103] events "msg"="Warning"  "message"="error deleting AzureMachine f0416348a3ff4c950718a2ed5690234e5f2e4a97/lin-wb5sp: failed to delete AzureMachine service virtualmachine: failed checking if the operation was complete: failed checking if the operation was complete: Code=\"OverconstrainedZonalAllocationRequest\" Message=\"Allocation failed. VM(s) with the following constraints cannot be allocated, because the condition is too restrictive. Please remove some constraints and try again. Constraints applied are:\\n  - Availability Zone\\n  - Preemptible VMs (VM might be preempted by another VM with a higher priority)\\n  - VM Size\\n\"" "object"={"kind":"AzureMachine","namespace":"f0416348a3ff4c950718a2ed5690234e5f2e4a97","name":"lin-wb5sp","uid":"6070578f-5857-4b79-bb35-4a08fe4e20ba","apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","resourceVersion":"162338100"} "reason"="Error deleting AzureMachine"
```

4. The `AzureMachine` is in a limbo state and capz-controller looping 4 delete requests per second (this seems a bit excessive) to the Azure Management API

**What did you expect to happen:**

The capz-controller successfully deletes the Azure resources related to the failed deployment.

**Anything else you would like to add:**

* Manually deleting the Azure VM, NIC and OS Disk does not resolve the issue - capz-controller still loops the same error message
* Manually deleting the resources in Azure and removing the `AzureMachine` finalizer removes the Kubernetes resource and ends the capz-controller loop

**Environment:**

- cluster-api-provider-azure version: 1.4.0
- Kubernetes version: (use `kubectl version`): 1.22.12
- OS (e.g. from `/etc/os-release`): Ubunut 20.04

**Comment by CecileRobertMichon** [comment]
@desek to clarify the repro steps, did you manually trigger a delete of the Machine and/or use a [MachineHealthCheck](https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/healthchecking.html) to delete the failed Machine? Or does the below indicate that the CAPZ controller started deleting a VM that was failing to create (which would be unexpected)?

> The capz-controller tries to delete the VM

**Comment by willie-yao** [comment]
@desek Do you have any further details on how to reproduce this issue like Cecile mentioned? I'm taking a look at the issue but I'm unable to reproduce the bug. Thanks in advance!

**Comment by desek** [comment]
We use MachineHealthCheck, yes. I'm assuming this is the one triggering the delete since we don't do it manually.

I can provide repro manifests in a couple of days. But in essence you can reproduce this by deploying Availability Zone pinned Spot instance VM with GPU in northeurope.

**Comment by CecileRobertMichon** [comment]
This was fixed by #2665 

/close
